### PR TITLE
Refinance

### PR DESCRIPTION
**Pull request by xurror**
Updated Bi monhtly scheduler to be more flexible and cover more cases.

### CODE REVIEW

**Comment by xurror** [comment]
@maektwain can you look at this and possibly merge too.
It changes you saw in the demo videos I sent you for the bi monthly scheduler.

**Comment by xurror** [comment]
@maektwain can you look at this and possibly merge too.
It changes you saw in the demo videos I sent you for the bi monthly scheduler.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/serialization/LoanProductDataValidator.java [hunk 1]
```diff
@@ -252,6 +253,33 @@ public void validateForCreate(final String json) {
...
+            baseDataValidator.reset().parameter(LoanProductConstants.SEMI_MONTH_START_DATE).value(firstSemiDate).notNull();
+
+            final LocalDate secondSemiDate = this.fromApiJsonHelper.extractLocalDateNamed(LoanProductConstants.SEMI_MONTH_SECOND_DATE,
+                    element);
+            baseDataValidator.reset().parameter(LoanProductConstants.SEMI_MONTH_SECOND_DATE).value(secondSemiDate).notNull()
+                    .validateDateAfter(firstSemiDate);
+
+            if (firstSemiDate.getMonth() != secondSemiDate.getMonth()) {
+                baseDataValidator.reset().parameter(LoanProductConstants.SEMI_MONTH_SECOND_DATE).failWithCode(
+                        "first.and.second.semi.dates.must.be.thesame.month", "First and second semi dates must be in the same month");
```
Comment (kevin-madhu) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanproduct/serialization/LoanProductDataValidator.java:268]: Is that missing "." between "the" and "same" a typo or intentional?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java [hunk 2]
```diff
@@ -70,51 +70,40 @@ public LocalDate generateNextRepaymentDate(final LocalDate lastRepaymentDate, fi
...
-                    int some = dueRepaymentPeriodDate.getDayOfMonth();
-                    int other = loanApplicationTerms.getFirstDateForSemi().getDayOfMonth();
-                    // diff
-                    int then = some - other;
-                    //
-                    if (then == 1) {
-                        dueRepaymentPeriodDate = dueRepaymentPeriodDate.minusDays(1);
-                    } else {
-                        dueRepaymentPeriodDate = dueRepaymentPeriodDate.minusDays(2);
+
```
Comment (kevin-madhu) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java:98]: Isn't this the same condition in line no. 79 above? So, can't this block be moved inside there?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java [hunk 3]
```diff
@@ -70,51 +70,40 @@ public LocalDate generateNextRepaymentDate(final LocalDate lastRepaymentDate, fi
...
+                // lastRepayment and dueRepaymentPeriodDate if they match semi monthly terms
+                if (lastRepaymentDate.getDayOfMonth() == loanApplicationTerms.getFirstDateForSemi().getDayOfMonth()) {
```
Comment (kevin-madhu) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/loanschedule/domain/DefaultScheduledDateGenerator.java:100]: Isn't this the same condition in line no. 79 above? So, can't this block be moved inside there?

**Review by kevin-madhu — APPROVED** [review]

**Pull request merged**